### PR TITLE
FIX: plots.image show labels for every row if provided

### LINKED ISSUE #4112: BUG: shap.plots.image no longer displays labels for rows other than the first one
### Issue Description

PR #3636 broke the display of labels on the second and subsequent rows in case:
- labels for each row are provided,
- no labels are provided but inferred from the Explainer object

Running the [Explain ResNet50 using the Partition explainer](https://shap.readthedocs.io/en/latest/example_notebooks/image_examples/image_classification/Explain%20ResNet50%20using%20the%20Partition%20explainer.html) example now no longer displays the labels on the second and subsequent rows.

### Minimal Reproducible Example

```python
# Almost straight from the example, unless marked otherwise

import json
from tensorflow.keras.applications.resnet50 import ResNet50, preprocess_input
import shap

model = ResNet50(weights="imagenet")
X, y = shap.datasets.imagenet50() 
X = X.astype(int) # Different from example to fix that actual images not showing correctly (unrelated issue)

url = "https://s3.amazonaws.com/deep-learning-models/image-models/imagenet_class_index.json"
with open(shap.datasets.cache(url)) as file:
    class_names = [v[1] for v in json.load(file).values()]

def f(x):
    tmp = x.copy()
    preprocess_input(tmp)
    return model(tmp)

masker = shap.maskers.Image("inpaint_telea", X[0].shape)
explainer = shap.Explainer(f, masker, output_names=class_names)

# Reduced number of evals for speed
# Reduced number of outputs shown to 3 (to have 2x3 instead of 2x4 so that one dimension is not a multiple of the other dimension)
shap_values = explainer(X[1:3], max_evals=20, batch_size=50, outputs=shap.Explanation.argsort.flip[:3])

# No labels on second row when running this:
shap.image_plot(shap_values)

# Same result when labels are provided
shap.image_plot(shap_values, labels=['a','b','c','d','e','f'])
shap.image_plot(shap_values, labels=[['a','b','c'],['d','e','f']])
```

### Traceback

```shell

```

### Expected Behavior

Labels should be displayed on the second row (and subsequent rows if more examples).

### Bug report checklist

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [x] I'd be interested in making a PR to fix this bug

### Installed Versions

```
shap==0.48.0
shap-0.48.1.dev2 (master)
```

### PR DESCRIPTION
**Pull request by julesvanrie**
## Overview

Closes #4112

Description of the changes proposed in this pull request:

`plots.image()` currently only displays labels for the first row, even if labels for every row (every image in the sample) have been provided. This breaks the use case in the [Explain ResNet50 using the Partition explainer](https://shap.readthedocs.io/en/latest/example_notebooks/image_examples/image_classification/Explain%20ResNet50%20using%20the%20Partition%20explainer.html) example. The use case is where the classes shown in the plot are not the same for every row (e.g. when one wants to only display the `top_k` classes, which would be different for every row).

This PR fixes this problem so that `plots.image()` displays:
- Labels for every row if they are provided.
- Only labels for the first row if only set of labels is provided.

This PR also adds a test case checking that `plots.image()` does not error when labels are provided in different formats that should be allowed to work.

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: tests/plots/test_image.py [hunk 1]
```diff
@@ -89,3 +89,49 @@ def __init__(self, data, values, output_names):
 
     shap_values_test = MockImageExplanation(test_data, test_values, test_output_names)
     shap.plots.image_to_text(shap_values_test)
+
+
+def test_image_plot_with_labels():
```
Comment (CloseChoice) [inline tests/plots/test_image.py]: Thanks for the test, this looks great. Would it be possible to set the seed for numpy and then use pytest-mpl to compare against the image?

Here is a blueprint: [test](https://github.com/shap/shap/blob/master/tests/plots/test_force.py#L94) and [baseline image](https://github.com/shap/shap/blob/master/tests/plots/baseline/test_force_plot_negative_sign.png).

**Review by CloseChoice — CHANGES_REQUESTED** [review]
Thanks for the fix, that looks great. Would be great if you could update the test, then we're good to merge.

**Comment by CloseChoice** [comment]
nevermind about the test failures, those are just connection issues when we try to download the dataset

File: tests/plots/test_image.py (hunk 1 above)
Comment (julesvanrie) [reply tests/plots/test_image.py]: Thanks for the review and the suggestion.

I have actually rewritten the tests to largely reuse the existing logic and example explanation of `test_image_multi` (by moving it into a fixture) and compare against baselines images that way.

**Review by CloseChoice — APPROVED** [review]
looks great, ready to merge when pipelines succeed! Thanks a lot, awesome work
